pkg/pquerier/frames.go

```diff
@@ -0,0 +1,660 @@
+
+	typedCol, ok := dc.data.([]float64)
+	if !ok {
+		return 0, fmt.Errorf("wrong type (type is %s)", dc.DType())
+	}
+
+	return typedCol[i], nil
+}
+
+// StringAt returns string value at index i
+func (dc *dataColumn) StringAt(i int) (string, error) {
+	if !dc.isValidIndex(i) {
+		return "", fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	typedCol, ok := dc.data.([]string)
+	if !ok {
+		return "", fmt.Errorf("wrong type (type is %s)", dc.DType())
+	}
+
+	return typedCol[i], nil
+}
+
+// TimeAt returns time.Time value at index i
+func (dc *dataColumn) TimeAt(i int) (int64, error) {
+	if !dc.isValidIndex(i) {
+		return 0, fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	typedCol, ok := dc.data.([]int64)
+	if !ok {
+		return 0, fmt.Errorf("wrong type (type is %s)", dc.DType())
+	}
+
+	return typedCol[i], nil
+}
+
+func (dc *dataColumn) SetData(d interface{}, size int) error {
+	dc.data = d
+	dc.size = size
+	return nil
+}
+
+func (dc *dataColumn) SetDataAt(i int, value interface{}) error {
+	if !dc.isValidIndex(i) {
+		return fmt.Errorf("index %d out of bounds [0:%d]", i, dc.size)
+	}
+
+	switch reflect.TypeOf(dc.data) {
+	case IntType:
+		dc.data.([]int64)[i] = value.(int64)
+	case FloatType:
+		// Update requested cell, only if not trying to override an existing value with NaN
+		if !(math.IsNaN(value.(float64)) && dc.data.([]float64)[i] != 0) {
+			dc.data.([]float64)[i] = value.(float64)
+		}
+	case StringType:
+		dc.data.([]string)[i] = value.(string)
+	case TimeType:
+		dc.data.([]time.Time)[i] = value.(time.Time)
+	case BoolType:
+		dc.data.([]bool)[i] = value.(bool)
+	}
+	return nil
+}
+
+func NewConcreteColumn(name string, colSpec columnMeta, size int, setFunc func(old, new interface{}) interface{}) *ConcreteColumn {
+	col := &ConcreteColumn{basicColumn: basicColumn{name: name, spec: colSpec, size: size}, setFunc: setFunc}
+	col.data = make([]interface{}, size)
+	return col
+}
+
+type ConcreteColumn struct {
+	basicColumn
+	setFunc func(old, new interface{}) interface{}
+	data    []interface{}
+}
+
+func (c *ConcreteColumn) DType() DType {
+	var a float64
+	return reflect.TypeOf(a)
+}
+func (c *ConcreteColumn) FloatAt(i int) (float64, error) {
+	if !c.isValidIndex(i) {
+		return 0, fmt.Errorf("index %d out of bounds [0:%d]", i, c.size)
+	}
+	if c.data[i] == nil {
+		return math.NaN(), nil
+	}
+	return c.data[i].(float64), nil
+}
+func (c *ConcreteColumn) StringAt(i int) (string, error) {
+	return "", errors.New("aggregated column does not support string type")
+}
+func (c *ConcreteColumn) TimeAt(i int) (int64, error) {
+	return 0, errors.New("aggregated column does not support time type")
+}
+func (c *ConcreteColumn) SetDataAt(i int, val interface{}) error {
+	if !c.isValidIndex(i) {
+		return fmt.Errorf("index %d out of bounds [0:%d]", i, c.size)
+	}
+
+	c.data[i] = c.setFunc(c.data[i], val)
+	return nil
+}
+func (c *ConcreteColumn) GetInterpolationFunction() (InterpolationFunction, int64) { return nil, 0 }
```

Comment:
consider implementing NOP function or return an error in case you don't want it to be used